server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -523,6 +525,28 @@ public String getHeader(String header) {
         return headers.get(header);
     }
 
+    /**
+     * This method adds a new header, <b>Only use this method if the header can not be passed from request headers</b>
+     * @param threadContext current thread context
+     */
+    public void addHeader(final String headerName, final ThreadContext threadContext, final Supplier<String> defaultValueSupplier) {
+        // For now this header will be coming from HTTP headers but in second phase this header
+
+        // We will use this constant from QueryGroup Service once the framework changes are done
+
+        String headerValue = threadContext.getHeader(headerName);
+
+        if (headerValue == null) {
+            headerValue = defaultValueSupplier.get();
+        }
+
+        final Map<String, String> newHeaders = new HashMap<>(headers);
+
+        newHeaders.put(headerName, headerValue);
+
+        this.headers = newHeaders;
+    }
```

Comment:
Can we introduce `QueryGroupAwareTask` for `SearchTask` and `SearchShardTask` instead of breaking the immutability of `Task`? For example:

```
public abstract class QueryGroupAwareTask extends CancellableTask {
  private String queryGroupId;

  public QueryGroupAwareTask(long id, String type, String action, String description, TaskId parentTaskId, Map<String, String> headers) {
    this(id, type, action, description, parentTaskId, headers);
  }

  public void setQueryGroupId();

  public String getQueryGroupId();
}
```

Reply:
Done.